### ISSUE TITLE
WebGL support in WK2 [QT5.9 / EGLFS QPA]

### ISSUE THREAD

**Issue by SylvainGarrigues** [annulen/webkit#536]
Hello Konstantin,

First, excellent work for this project.

While WebGL works in WK1 (see example at the end), it doesn't work on WK2 (see example below). Here is my test URL:
http://www.equasys.de/solarsystem.html

### My QML file:
```
import QtWebKit 3.0
import QtQuick 2.6
import QtQuick.Controls 1.5
import QtQuick.Controls.Styles 1.4
import QtQuick.Dialogs 1.2
import QtQuick.Layouts 1.3

ApplicationWindow {
    width: 500
    visible: true
    title: "Annulen"
	ScrollView {
         anchors.fill: parent

	WebView {
         anchors.fill: parent
         url: "http://www.equasys.de/solarsystem.html"
        }
    }
}
```

### My C++ file:
```
#include <QApplication>
#include <QQmlApplicationEngine>

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
	QApplication::setAttribute(Qt::AA_ShareOpenGLContexts, true);


    QQmlApplicationEngine engine;
    engine.load(QUrl(QStringLiteral("qrc:/main.qml")));

    return app.exec();
}
```

###However this works (WK1 right?):
#include <QApplication>
#include <QWebView>
#include <QWebInspector>

int main(int argc, char** argv) {
    QApplication app(argc, argv);
    QWebView view;

    view.setUrl(QUrl("http://madebyevan.com/webgl-water/"));
    view.show();
    return app.exec();
}

**Comment by annulen** [comment]
Same issues #525 probably

**Comment by annulen** [comment]
BTW, AFAIK WebGL never worked on EGLFS because there is no standard way to share OpenGL textures between processes without GLX

**Comment by SylvainGarrigues** [comment]
http://thebugfreeblog.blogspot.fr/2015/12/hardware-acceleration-on-webkit-1.html

He does manage to get it with old WebKit.

**Comment by annulen** [comment]
"I'm not talking about WebGL here"

It may be a good idea to talk to luc4 on our IRC channel, maybe he can provide further info

**Comment by annulen** [comment]
What is your hardware platform?

**Comment by SylvainGarrigues** [comment]
RPI2, FreeBSD, QT5.9 backend. Everything working perfectly except WebGL on WK2 :-)

**Comment by SylvainGarrigues** [comment]
I've seen this: https://bugreports.qt.io/browse/QTBUG-33413 where one claimed he had patches for WebGL functionality to WebKit2.

I also read other Carl's blog posts where a commenter mentioned Qt::AA_ShareOpenGLContexts flag had to be used on QGuiApplication.

**Comment by SylvainGarrigues** [comment]
> BTW, AFAIK WebGL never worked on EGLFS because there is no standard way to share OpenGL textures between processes without GLX

You mean on WK2 right? Because I can show you a video of WebGL working on EGLFS with your webkit fork and WK1 :-)

**Comment by annulen** [comment]
>You mean on WK2 right?

Right. In WK1 page and GUI is in one process, so no sharing needed.

AFAIU, Qt::AA_ShareOpenGLContexts is used to share context between different top-level windows in the same processes, but cannot help with anything for cross-process sharing.

To get cross-process sharing working there are 3 ways:
* Use GLX (X11). It's not yet possible in my fork before  #525 is fixed, however works in legacy QtWebKit
* Use Wayland. There is (quite new) implementation for GTK port at https://bugs.webkit.org/show_bug.cgi?id=115803, which AFAIU implements nested compositor inside UI process for which WebProcess conects. May require substantial adaptations, and probably better left for "webkit upgrade" milestone where we at least don't need to merge that patch anymore
* Use platform-specific API. For example, Broadcom provides their own compositor API as a part of their Nexus driver, however I'm pretty sure it's not available for FreeBSD

**Comment by SylvainGarrigues** [comment]
Thanks. Is it worth trying USE_COORDINATED_GRAPHICS_THREADED (instead of the default PROCESS one ?)

**Comment by annulen** [comment]
It won't work with our current implementation of WK2. It maybe worth switching to it instead of the current model (esp. given that it was purged from trunk when EFL port left it recently), but it's a research topic.
